pkg/cli/tail.go

```diff
@@ -222,24 +267,67 @@ func (c *TailCommand) Run(ctx context.Context, args []string) error {
 	return nil
 }
 
-func (c *TailCommand) tail(ctx context.Context) ([]*loggingpb.LogEntry, error) {
-	var p logPuller
-	if c.testPuller != nil {
-		p = c.testPuller
-	} else {
-		logClient, err := logging.NewClient(ctx)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create logging client: %w", err)
-		}
-		p = cloudlogging.NewPuller(ctx, logClient, c.flagScope)
-	}
+func (c *TailCommand) streamTail(ctx context.Context, extra []validation.Validator, puller logPuller) error {
+	// logCh is used to receive logEntriesfrom StreamPull
+	logCh := make(chan *loggingpb.LogEntry)
+	// errCh is used to receive error from StreamPull
+	errCh := make(chan error)
+	// doneCh is used to notifiy the main goroutinue when
+	// StreamPull is finished or canceled.
+	doneCh := make(chan struct{})
 
-	ls, err := p.Pull(ctx, c.queryFilter(), c.flagMaxNum)
-	if err != nil {
-		return nil, fmt.Errorf("failed to pull logs: %w", err)
-	}
+	var failCount int
+	var totalCount int
 
-	return ls, nil
+	go func() {
+		if err := puller.StreamPull(ctx, c.queryFilter(), logCh); err != nil {
+			errCh <- fmt.Errorf("StreamPull failed: %w", err)
+		}
+		close(logCh)
+		close(errCh)
+	}()
+
+	go func() {
+		// close doneCh to notify main go rountine
+		// that receiving from StreamPull has finished.
+		defer func() {
+			if c.flagValidate {
+				c.Outf("Validation failed for %d logs (out of %d)", failCount, totalCount)
+			}
+			close(doneCh)
+		}()
+
+		for l := range logCh {
+			totalCount++
+			js, err := protojson.Marshal(l)
+			if err != nil {
+				c.Errf("failed to marshal log to json (InsertId: %q): %w", l.InsertId, err)
+				continue
+			}
+			c.Outf(stripSpaces(string(js)))
+			if c.flagValidate {
+				if err := validation.Validate(string(js), extra...); err != nil {
+					failCount++
+					c.Errf("failed to validate log (InsertId: %q): %w\n", l.InsertId, err)
+				} else {
+					c.Outf("Successfully validated log (InsertId: %q)\n", l.InsertId)
+				}
+			}
+		}
+	}()
+
+	select {
+	case <-ctx.Done():
```

Comment:
Why do we care about this? Context is cancelled => StreamPull will return => close the `logCh` => exit the for loop on `logCh` => close the `doneCh`. So IIUC, you can just wait for `<-doneCh`?

Reply:
Updatad, removed errCh, and just wait on doneCh.

Reply:
Made another update, now doneCh is also removed.